toolchain/check/cpp_thunk.cpp

```diff
@@ -0,0 +1,369 @@
+
+      llvm::APInt bit_width =
+          context.ints().Get(context.types().GetIntTypeInfo(type_id).bit_width);
+      return bit_width != 32 && bit_width != 64;
+    }
+
+    default:
+      return true;
+  }
+}
+
+auto IsCppThunkRequired(Context& context, const SemIR::Function& function)
+    -> bool {
+  if (!function.clang_decl_id.has_value()) {
+    return false;
+  }
+
+  if (function.self_param_id.has_value()) {
+    // TODO: Support member methods.
+    return false;
+  }
+
+  SemIR::TypeId return_type_id =
+      function.GetDeclaredReturnType(context.sem_ir());
+  if (return_type_id.has_value()) {
+    // TODO: Support non-void return values.
+    return false;
+  }
+
+  bool thunk_required_for_param = false;
+  for (auto param_id :
+       context.inst_blocks().GetOrEmpty(function.call_params_id)) {
+    if (param_id == SemIR::ErrorInst::InstId) {
+      return false;
+    }
+    if (!thunk_required_for_param &&
+        IsThunkRequiredForType(
+            context,
+            context.insts().GetAs<SemIR::AnyParam>(param_id).type_id)) {
+      thunk_required_for_param = true;
+    }
+  }
+
+  return thunk_required_for_param;
+}
+
+// Returns whether the type is a pointer or a signed int of 32 or 64 bits.
+static auto IsScalarType(clang::ASTContext& ast_context, clang::QualType type)
+    -> bool {
+  if (type->isPointerType()) {
+    return true;
+  }
+
+  if (const auto* builtin_type = type->getAs<clang::BuiltinType>()) {
+    // TODO: Add support for other integer types.
+    if (builtin_type->getKind() == clang::BuiltinType::Int) {
+      uint64_t type_size = ast_context.getTypeSize(type);
+      return type_size == 32 || type_size == 64;
+    }
+  }
+
+  return false;
+}
+
+// Creates the thunk parameter types given the callee function. Also returns for
+// each type whether it is different from the matching callee function parameter
+// type.
+static auto BuildThunkParameterTypes(
+    clang::ASTContext& ast_context,
+    const clang::FunctionDecl& callee_function_decl)
+    -> std::tuple<llvm::SmallVector<clang::QualType>,
+                  llvm::SmallVector<bool> > {
+  unsigned num_params = callee_function_decl.getNumParams();
+  llvm::SmallVector<clang::QualType> thunk_param_types;
+  thunk_param_types.reserve(num_params);
+  llvm::SmallVector<bool> param_type_changed;
+  param_type_changed.reserve(num_params);
+
+  for (const clang::ParmVarDecl* callee_param :
+       callee_function_decl.parameters()) {
+    clang::QualType param_type = callee_param->getType();
+    bool is_scalar_type = IsScalarType(ast_context, param_type);
+    if (!is_scalar_type) {
+      clang::QualType pointer_type = ast_context.getPointerType(param_type);
+      param_type = ast_context.getAttributedType(
+          clang::NullabilityKind::NonNull, pointer_type, pointer_type);
+    }
+    param_type_changed.push_back(!is_scalar_type);
+    thunk_param_types.push_back(param_type);
+  }
+
+  return {thunk_param_types, param_type_changed};
```

Comment:
This will perform `SmallVector` copies; it'd be better to use `std::move` here, but even then you'd copy the small buffer of the `SmallVector`. Suggested edits would avoid this.
```suggestion
  return result;
```

Reply:
Done.